### ISSUE TITLE
Using single quotes around strings that don't require expansion/evaluation

### ISSUE THREAD

**Issue by Mikecom32** [PoshCode/PowerShellPracticeAndStyle#63]
I searched and didn't see a previous issue for this. Do you think we should recommend using single quotes around any string that doesn't require some type of evaluation?

I'd propose that if a string is just a string, it should be surrounded by single quotes:

    $testVariable1 = 'This is a plain string.'
    $testVariable2 = 'This is a string containing a $variable I don't want evaluated'
    $testVariable3 = "This is a string containing a $variable I want evaluated"

as opposed to:

    $testVariable1 = "This is a plain string."
    $testVariable2 = 'This is a string containing a $variable I don't want evaluated'
    $testVariable3 = "This is a string containing a $variable I want evaluated"

I think this improves readability and is more consistent. Also, while the difference is very minimal, single quoted strings generally perform faster than double quoted strings when there is no expansion/evaluation.

A bit more info:
- http://ramblingcookiemonster.github.io/Single-Quotes-or-Double-Quotes/
- http://jon.netdork.net/2015/09/08/powershell-and-single-vs-double-quotes/

**Comment by dlwyatt** [comment]
I don't really see any difference in readability.  It does help to avoid bugs on occasion, though.  I use this practice myself, but I don't stress about it if I'm working on a project where someone else has double-quoted string literals without any expansion.

**Comment by Jaykul** [comment]
The performance argument is simply incorrect.

    Scenario         AverageSeconds 
    --------         -------------- 
    Single-10,000             0.017 
    Double-10,000             0.015 
    Single-1,000,000          1.403 
    Double-1,000,000          1.391

The argument about readability not convincing. Especially when your example contains a perfect counter-example: the word "don't" ... your code is broken.

Given that double-quoted strings are faster, and that readability in this case is in the eye of the beholder, I don't see any reason to make a language-wide best practice recommendation here. As always, individual authors, companies, or projects may want to be more strict and take a stance one way or the other.

One might, for instance, want to **forbid** single-quoted strings and require using double-quotes for everything, with escapes where necessary -- to avoid switching back and forth, and prevent problems like "don't" which crop up when people use things that are unfamiliar.

Alternatively, one might simply forbid using expressions in strings, and require single-quoted strings everywhere, and the use of string formatting to embed calculated values (like C# prior to 6.0).

**Comment by rkeithhill** [comment]
I think there is a "bit" of readability, or perhaps scan-ability, improvement when using a sq string.  If I see one of those I know I can quickly scan past it, confident I don't have to look for any subexpression script.  

That said, for something like `don't` - I'd probably switch to a dq string.  I do the same thing with XML attribute values which accept both sq and dq strings.  I use whichever is most convenient if there are quote chars in the string.

**Comment by mattmcnabb** [comment]
I like Joel's point about maintaining a script down the road. I've probably lost a lot of time writing scripts when I need to go back and change single quotes to double quotes in order to include a variable expansion. If all strings were simply double-quoted then I could shave some time off of script maintenance.

I've switched to this practice recently, but for your own work you may need to weigh it's benefits against the possibility of introducing a bug. Of course this *should* be prevented by using a good editor with the right syntax highlighting.

**Comment by Mikecom32** [comment]
@Jaykul The performance reason was something I was quoting from the two blogs I linked above. I haven't tested it myself. Either way, it's incredibly minor.

To be fair, the issue with quotes in strings crop up with both conventions. It's pretty funny that I screwed it up myself freehanding into the github text box.

I think it is/was worth discussing at least.

**Comment by Jaykul** [comment]
Yeah, it's definitely a question that comes up, and the difference between single and double and here-strings is something we should make sure new scripters understand. I don't think we can make a recommendation that a certain way is definitely better, but this conversation about it is definitely worth having and keeping a log of 😀

**Comment by RamblingCookieMonster** [comment]
I'm all for espousing the benefits and caveats (slight performance impact even at massive scale, subjective readability), but I'm not sure if this fits as a *good* practice, let alone a *best* practice.

On the other side, I can't tell you how much time I've wasted using single quotes and having to go back and switch to double quotes.

Cheers!
